src/batcher/solve_pair.rs

```diff
@@ -1,16 +1,359 @@
 use crate::models::order::Order;
 use crate::models::solution::Solution;
 use anyhow::Result;
-use web3::types::U256;
+use web3::types::{U256, H256, Address};
+
+
+trait Matchable {
+    /// Returns whether there is a price that satisfies both orders.
+    fn have_price_overlap(&self, other: &Order) -> bool;
+}
+
+
+impl Matchable for Order {
+    fn have_price_overlap(&self, other: &Order) -> bool {
+        self.sell_amount > U256::zero()
+            && other.sell_amount > U256::zero()
+            && self.buy_amount * other.buy_amount <= self.sell_amount * other.sell_amount
+    }
+}
+
 
 pub fn solve_pair(
     sell_orders_token0: &Vec<Order>,
     sell_orders_token1: &Vec<Order>,
 ) -> Result<Solution> {
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token0));
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token1));
+    
+    // Get number of orders in each direction.
+    let nr_orders_token0 = sell_orders_token0.len();
+    let nr_orders_token1 = sell_orders_token1.len();
+
+    // Init vectors of sell amounts.
+    let mut executed_sell_orders_token0: Vec<Order> = vec![];
+    let mut executed_sell_orders_token1: Vec<Order> = vec![];
+
+    // Match orders with best limit prices, if possible.
+    if !(sell_orders_token0.is_empty() || sell_orders_token1.is_empty()) {
+
+        // The best orders are the last elements in their vectors.
+        let best_sell_order_token0 = &sell_orders_token0[nr_orders_token0 - 1];
```

Comment:
You can also use the pop function:
See here:
https://doc.rust-lang.org/std/vec/struct.Vec.html
in order to avoid the nr_order_token declarations

Reply:
`pop` requires a mutable vec, but there is also [`last`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.last). Then the if statement can be re-written as:

```rust
if let (Some(best_sell_order_token0), Some(best_sell_order_token1)) = (sell_orders_token0.last(), sell_orders_token1.last()) {

}
```

Reply:
That's nice, thanks! Changed it in ac676c0